services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -5399,6 +5388,20 @@ private OfflinePushStatusInfo getOfflinePushStatusInfo(
       Store store,
       int versionNumber) {
     Pair<ExecutionStatus, String> statusAndDetails;
+
+    HelixVeniceClusterResources resources = getHelixVeniceClusterResources(clusterName);
+    MaintenanceSignal maintenanceSignal =
+        HelixUtils.getClusterMaintenanceSignal(clusterName, resources.getHelixManager());
+
+    // Check if cluster is in maintenance mode due to too many offline instances
+    if (maintenanceSignal != null && maintenanceSignal
+        .getAutoTriggerReason() == MaintenanceSignal.AutoTriggerReason.MAX_OFFLINE_INSTANCES_EXCEEDED) {
+      String msg = "Push errored out for " + kafkaTopic + "due to too many offline instances. Reason: "
+          + maintenanceSignal.getReason();
+      LOGGER.error(msg);
+      return new OfflinePushStatusInfo(ExecutionStatus.ERROR, msg);
```

Comment:
IIUC, this will affect all pushes even though they are not onboarded to colo-by-colo push, right?

Where do we plan to do the cluster health and store health validation? In controller or VPJ?

Reply:
Yes right, it will trigger regardless of the colo-by-colo push feature. But this will help in stopping further bad pushes if canary push creates bad health. Currently all the logics for validations are in controller.